Methods for independent sets and cliques

nolberg opened the pull request:
See pull request #248.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/Clique.java:
```diff
@@ -0,0 +1,88 @@
+package org.jgrapht.alg;
```
All files need standard project headers.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/Clique.java:
```diff
@@ -0,0 +1,88 @@
+
+public abstract class Clique {
+	
+	/**
```
Please set your editor to use spaces instead of tabs for indentation.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/Clique.java:
```diff
@@ -0,0 +1,88 @@
+ * @author Nils Olberg
+ */
+
+public abstract class Clique {
```
There's already one clique finder implementation (BronKerboschCliqueFinder), and now this is a second one.  So we should try to rationalize the interface/package/naming.

The pattern we've started following is to add an interface in the interfaces package, then create a new subpackage (e.g. clique) to contain the implementation classes (and "moving" the existing one via deprecation).

So it would be something like interface CliqueFinder, with implementations BronKerboschCFImpl and IndependentSetComplementCFImpl.

@jkinable what do you think?

jsichi replied:
(The new interfaces introduced may need to be more refined based on differences in the problems being solved by the different implementations.)

jkinable replied:
Similar to the MinimumVertexCoverAlgorithm and MinimumWeightedVertexCoverAlgorithm which solves the mimimum (weighted) vertex cover problemI would add:
1. MaximumCliqueAlgorithm and MaximumWeightedCliqueAlgorithm which solve the maximum (weigthed) clique problem.
2. MaximumIndependentSetAlgorithm and  MaximumWeightedIndependentSetAlgorithm

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/IndependentSet.java:
```diff
@@ -0,0 +1,73 @@
+ * @author Nils Olberg
+ */
+
+public abstract class IndependentSet {
```
This is just one of the possible ways to find independent sets, so we may end up in a similar situation (needing to split interface from implementation).

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/Clique.java:
```diff
@@ -0,0 +1,88 @@
+     * @param weights			map with non-negative weights for vertices
+     * @return weight of maximum weight clique
+     */
+	public static <V, E> Double getWeight(UndirectedGraph<V, E> graph, Map<V, Double> weights) 
```
A common use-case is: get the maximum (weighted) clique and its corresponding weight. This current implementation requires 2 invocations of an expensive algorithm to find this solution.

jkinable commented:
Here's my recommendation: have a look at the MinimumVertexCoverAlgorithm, MinimumWeightedVertexCoverAlgorithm interfaces, mimic that for the clique/independent sets. Next, remove the static methods and replace them by normal methods, similar to what is done for the various vertex cover algorithm implementations.

nolberg commented:
I made the following changes:

1. Created interfaces for (weighted) clique and independent set algorithms 
2. Added classes that implement these interfaces
3. Replaced the static by normal methods. 

It looks very similar to the vertex cover case. I did not modify BronKerboschCliqueFinder since it is used to compute collections of cliques and not just a single clique.

nolberg commented:
In order to reduce the amount of redundant code, I thought it makes sense to remove the interfaces VertexCover, Clique and IndependentSet together with the corresponding implementations. Now there is just one class WeightedVertexSet, which stores a set of vertices and a weight associated with the set.

d-michail commented:
@nolberg I think it is more intuitive for a user to have the different interfaces: VertexCover, Clique and IndependentSet. 

@jkinable I would also remove the VertexCoverImpl from inside the MinimumVertexCoverAlgorithm as it is not an interface but a default implementation. The best location for it would be inside the org.jgrapht.alg.vertexcover package. Maybe even call it DefaultVertexCoverImpl.

Sabraine commented on jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/MaximumWeightedIndependentSetAlgorithm.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    default WeightedVertexSet<V> getIndependentSet(UndirectedGraph<V,E> graph){
+        Map<V, Double> vertexWeightMap = graph.vertexSet().stream().collect(Collectors.toMap(Function.identity() , vertex-> 1.0));
```
I have error "illegal start of expression"  and cannot find symbol "vertex"! can you help me please?

jsichi commented:
Closing this one out since we didn't reach consensus on the interfaces.  See this doc:

https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor#plan-b
